File: cpp/cpp-pthread/demo18a03-barrier.cpp
/*
BARRIERS AND LATCHES
Version A: Cyclic barriers
*/


#include <iostream>
#include <string>
#include <tuple>
#include <unistd.h>
#include <pthread.h>
using namespace std;



pthread_barrier_t syncPointA;
pthread_barrier_t syncPointB;



void* processRequest(void* argVoid) {
    auto arg = *(tuple<string,int>*) argVoid;
    string userName = std::get<0>(arg);
    int waitTime = std::get<1>(arg);

    sleep(waitTime);

    cout << "Get request from " << userName << endl;
    pthread_barrier_wait(&syncPointA);

    cout << "Process request for " << userName << endl;
    pthread_barrier_wait(&syncPointB);

    cout << "Done " << userName << endl;

    pthread_exit(nullptr);
    return nullptr;
}



int main() {
    constexpr int NUM_THREADS = 4;
    pthread_t lstTid[NUM_THREADS];
    int ret = 0;

    // tuple<userName, waitTime>
    tuple<string,int> lstArg[NUM_THREADS] = {
        { "lorem", 1 },
        { "ipsum", 3 },
        { "dolor", 3 },
        { "amet", 10 }
    };

    ret = pthread_barrier_init(&syncPointA, nullptr, 2);
    ret = pthread_barrier_init(&syncPointB, nullptr, 2);

    for (int i = 0; i < NUM_THREADS; ++i) {
        ret = pthread_create(&lstTid[i], nullptr, &processRequest, &lstArg[i]);
    }

    for (auto&& tid : lstTid) {
        ret = pthread_join(tid, nullptr);
    }

    ret = pthread_barrier_destroy(&syncPointA);
    ret = pthread_barrier_destroy(&syncPointB);
    return 0;
}

File: python/demo18a03_barrier.py
'''
BARRIERS AND LATCHES
Version A: Barriers
'''

import time
import threading



sync_point_a = threading.Barrier(parties=2)
sync_point_b = threading.Barrier(parties=2)



def process_request(user_name: str, wait_time: int):
    time.sleep(wait_time)

    print(f'Get request from {user_name}')
    sync_point_a.wait()

    print(f'Process request for {user_name}')
    sync_point_b.wait()

    print(f'Done {user_name}')



lstarg = [
    ('lorem', 1),
    ('ipsum', 3),
    ('dolor', 3),
    ('amet', 10)
]

_ = [ threading.Thread(target=process_request, args=arg) for arg in lstarg ]
